WiiFlow support for autoboot games via arguments in WiiSX 4.0 (Wiimpathy)

saulfabregwiivc opened the pull request:
With WiiSX Beta 4.0 (https://github.com/emukidid/pcsxgc/releases/tag/Beta_4.0), my old PR for add WiiFlow support (https://github.com/emukidid/pcsxgc/pull/10) is now obsolete.

Basing on latest WiiSX 4.0, i decided to try again to add WiiFlow support, for autolaunch PlayStation 1 games automatically at startup when passing two arguments, first one with the game path and the second one with the game file to load.

Usage in meta.xml:
```
<arguments>
<arg>sd:/wiisx/isos</arg>
<arg>CrashBandicoot.cue</arg>
</arguments>
```

Hope this can be merged soon :D

saulfabregwiivc committed:
WiiFlow support for autoboot games via arguments in WiiSX 4.0 (Wiimpathy)

pcercuei commented on Gamecube/GamecubeMain.cpp:
```diff
@@ -342,10 +342,26 @@ void video_mode_init(GXRModeObj *videomode, u32 *fb1, u32 *fb2, u32 *fb3)
 {
 	/* INITIALIZE */
 #ifdef HW_RVL
+	if(argc > 2 && argv[1] != NULL && argv[2] != NULL)
```
Why do you need to check against NULL? Is that even possible?

saulfabregwiivc replied:
from what i understood in this piece of code check if the first argument is a valid path which exists and check if the second argument is a valid file name which exists.
If these are not valid, they will throw NULL and the emulator will start displaying the menu.

pcercuei commented on Gamecube/GamecubeMain.cpp:
```diff
@@ -342,10 +342,26 @@ void video_mode_init(GXRModeObj *videomode, u32 *fb1, u32 *fb2, u32 *fb3)
+	{
+		Autoboot = false;
+		memset(AutobootPath, 0, sizeof(AutobootPath));
+		memset(AutobootROM, 0, sizeof(AutobootROM));
```
These are the default, the whole "else" case can be dropped.

saulfabregwiivc replied:
i'm not sure about deleting the whole else.
We need also to tell the emulator if we are not autobooting games, then empty the autoboot variables and say Autoboot = false.

pcercuei replied:
If you are not sure, then maybe it's time you learn C++.

pcercuei commented on Gamecube/menu/MenuContext.cpp:
```diff
@@ -160,3 +163,13 @@ void MenuContext::draw()
+
+void MenuContext::Autoboot()
+{
+	if(strcasestr(AutobootPath,"sd:/") != NULL)
```
You can pass down the argv[1] / argv[2] variables as arguments to the Autoboot() function, and Func_LoadFromSD / Func_LoadFromUSB(), then you don't need to use static buffers.

pcercuei commented on Gamecube/menu/LoadRomFrame.cpp:
```diff
@@ -175,8 +183,16 @@ void Func_LoadFromUSB()
+	else
+	{
+		pMenuContext->setActiveFrame(MenuContext::FRAME_FILEBROWSER,loadRomMode);
+		fileBrowserFrame_OpenDirectory(isoFile_topLevel);
```
`fileBrowserFrame_OpenDirectory` is called in both branches of the if(), so it can be factorized.

The whole if() can be factorized as well since it's duplicated between `Func_LoadFromSD` and `Func_LoadFromUSB`.

pcercuei commented on Gamecube/menu/MenuContext.cpp:
```diff
@@ -55,7 +55,10 @@ MenuContext::MenuContext(GXRModeObj *vmode)
-	setActiveFrame(FRAME_MAIN);
+	
+	// Don't show the menu if we autoboot a game.
+	if (!strlen(&AutobootROM[0]))
```
```c
if (!AutobootROM[0]) { ... }
```

saulfabregwiivc replied:
i would try it later and tell the results.

pcercuei commented on Gamecube/menu/FileBrowserFrame.cpp:
```diff
@@ -472,6 +473,17 @@ void fileBrowserFrame_LoadFile(int i)
+				// Wiimpathy:
+				// FIXME: The MessageBox is a hacky way to fix input not responding.
+				// No time to improve this...
+				menu::MessageBox::getInstance().setMessage("Autobooting game...");
```
This is extremely hacky and I don't like it. Just add a new method in the MessageBox class, why do you need to do it in setMessage() at all?

saulfabregwiivc replied:
https://github.com/emukidid/pcsxgc/pull/32/commits/f7ae144eb86236038af18a10f0c5d40113623457 removes this hacky way and replaces it with a sleep command instead.

saulfabregwiivc committed:
remove hacky MessageBox, replace it with a sleep command

Instead of making a hacky new message box, we replaced it with waiting for 5 (milisecs?) after init WPAD for avoid input issues when in autooot.

Original code by FIX94.

saulfabregwiivc committed:
forgot to include unistd for sleep() function

saulfabregwiivc committed:
simplify detection of autoboot, part 1

Suggested by pcercuei

saulfabregwiivc committed:
Merge branch 'master' into wiisx-4.0-wiiflow

toxic9 commented:
Hello. Can anyone fix and implement this feature, please?
It's very important so it can be used as plugin.

I tried to make some changes and compile it, but I'm having so much trouble with multiple compilation errors.
Thanks

emukidid commented:
Happy to dust this off and implement it based on the original work done here. I'll add it to my list!

emukidid commented:
I ended up writing it completely different in the end.

@toxic9 can you test this build out? https://github.com/emukidid/pcsxgc/actions/runs/25254122251/artifacts/6763627609

Here's the plugin config I used for it:
```
[PLUGIN]
arguments={device}:/{path}/{name}
bannersound=ZZ--Sounds/PS1.ogg
consolecoverid=Sony
covercolor=111111
coverfolder=PlayStation
displayname=PlayStation (WiiSX)
dolfile=PlayStation/WiiSX/WiiSX.dol
filetypes=.cue|.img|.iso|.cdi|.chd
magic=57323434
romdir=wiisx/isos
rompartition=-1
```
